rdflib/plugins/sparql/evaluate.py

```diff
@@ -381,18 +381,22 @@ def _yieldBindingsFromServiceCallResult(ctx: QueryContext, r, variables):
     res_dict: Dict[Variable, Identifier] = {}
     for var in variables:
         if var in r and r[var]:
-            if r[var]["type"] == "uri":
-                res_dict[Variable(var)] = URIRef(r[var]["value"])
-            elif r[var]["type"] == "bnode":
-                res_dict[Variable(var)] = BNode(r[var]["value"])
-            elif r[var]["type"] == "literal" and "datatype" in r[var]:
+            d = r[var]
+            t = d["type"]
+            if t == "uri":
+                res_dict[Variable(var)] = URIRef(d["value"])
+            elif t == "literal":
                 res_dict[Variable(var)] = Literal(
-                    r[var]["value"], datatype=r[var]["datatype"]
+                    d["value"], datatype=d.get("datatype"), lang=d.get("xml:lang")
                 )
-            elif r[var]["type"] == "literal" and "xml:lang" in r[var]:
+            elif t == "typed-literal":
                 res_dict[Variable(var)] = Literal(
-                    r[var]["value"], lang=r[var]["xml:lang"]
+                    d["value"], datatype=URIRef(d["datatype"])
                 )
+            elif t == "bnode":
+                res_dict[Variable(var)] = BNode(d["value"])
+            else:
+                raise NotImplementedError("json term type %r" % t)
```

Comment:
I know there is a lot of places where this is used like this, but I think `NotImplementedError` is not appropriate in this case as it is described as

https://docs.python.org/3/library/exceptions.html#NotImplementedError

> This exception is derived from [RuntimeError](https://docs.python.org/3/library/exceptions.html#RuntimeError). In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

I would rather go for [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError):
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError).

```suggestion
                raise ValueError(f"invalid type {t!r} for variable binding {var!r}")
```

Don't apply this suggestion though, I will made a commit on your branch to fix this and some other issues.

Reply:
Fixed on your branch now.